lightning/src/ln/channel.rs

```diff
@@ -3822,33 +4070,77 @@ pub(crate) fn get_legacy_default_holder_selected_channel_reserve_satoshis(channe
 ///
 /// This is used both for outbound and inbound channels and has lower bound
 /// of `dust_limit_satoshis`.
-#[cfg(any(dual_funding, splicing))]
 fn get_v2_channel_reserve_satoshis(channel_value_satoshis: u64, dust_limit_satoshis: u64) -> u64 {
 	// Fixed at 1% of channel value by spec.
 	let (q, _) = channel_value_satoshis.overflowing_div(100);
 	cmp::min(channel_value_satoshis, cmp::max(q, dust_limit_satoshis))
 }
 
+pub(super) fn calculate_our_funding_satoshis(
+	is_initiator: bool, funding_inputs: &[(TxIn, TransactionU16LenLimited)],
+	funding_feerate_sat_per_1000_weight: u32, holder_dust_limit_satoshis: u64,
+) -> Result<u64, APIError> {
+	let mut total_input_satoshis = 0u64;
+	let mut our_contributed_weight = 0u64;
+
+	for (idx, input) in funding_inputs.iter().enumerate() {
+		if let Some(output) = input.1.as_transaction().output.get(input.0.previous_output.vout as usize) {
+			total_input_satoshis = total_input_satoshis.saturating_add(output.value.to_sat());
+			our_contributed_weight = our_contributed_weight.saturating_add(estimate_input_weight(output).to_wu());
```

Comment:
Sorry for missing this comment which is a super important one :man_facepalming: 

It may be possible if they provide "dummy" witnesses or provide the sizes for the witnesses.

They won't be able to provide valid witnesses (that involve signatures) before construction of the funding transaction is finished.

The issue is since this transaction is always using SIGHASH_ALL for any inputs sigs, we don't know what the sighash we need to sign will be for the funding tx at this point as we have no idea of the inputs or outputs the counterparty will provide (except for the 2-of-2 funding output of course). So the signatures cannot be created by the user before construction is done. At best we can estimate based on the output type for each input provided by the user. 

Thoughts? Or has Fry-day fried my brain? I'm actually hoping that's the case and it's a super trivial fix :P

Reply:
Yeah we don't care about the weight of the counterparty's inputs/outputs, and we could ask the user to provide their own witness weights or dummy witnesses that use dummy signatures where appropriate.

What they won't be able to provide is *valid* signatures in witnesses until after the whole tx is constructed as it's SIGHASH_ALL, but we don't need them yet, just the sizes.

We will have to ask the user again to provide those valid witnesses after transaction construction like we do are doing for V2.